### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by jforissier**
Removing more duplicate code in plat-*.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
```
Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>
```

**Comment by jforissier** [comment]
Addressed review comments.

File: core/arch/arm/plat-stm/conf.mk [hunk 1]
```diff
@@ -8,7 +8,7 @@ CFG_ARM32_core ?= y
 core-platform-cppflags	 = -I$(arch-dir)/include
 
 core-platform-subdirs += \
-	$(addprefix $(arch-dir)/, kernel mm sm tee sta) $(platform-dir)
+        $(addprefix $(arch-dir)/, kernel mm sm tee sta) $(platform-dir)
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-stm/conf.mk]: Space instead of tabs

File: core/arch/arm/tee/entry.c [hunk 2]
```diff
@@ -386,9 +386,59 @@ static void tee_entry_call_with_arg(struct thread_smc_args *args)
...
+	case TEESMC_OPTEE_L2CC_MUTEX_ENABLE:
+		ret = tee_l2cc_mutex_configure(
+			SERVICEID_ENABLE_L2CC_MUTEX, NULL);
+		break;
+	case TEESMC_OPTEE_L2CC_MUTEX_DISABLE:
+		ret = tee_l2cc_mutex_configure(
+			SERVICEID_DISABLE_L2CC_MUTEX, NULL);
+		break;
+	default:
+		args->a0 = TEESMC_RETURN_EBADCMD;
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/tee/entry.c]: ret not initialized in this case

File: core/arch/arm/plat-sunxi/main.c [hunk 3]
```diff
@@ -190,41 +163,8 @@ static void main_tee_entry(struct thread_smc_args *args)
...
-	args->a2 = TEESMC_OS_OPTEE_UUID_R2;
-	args->a3 = TEESMC_OS_OPTEE_UUID_R3;
-}
-
-/* Override weak function in tee/entry.c */
-void tee_entry_get_os_revision(struct thread_smc_args *args)
-{
-	args->a0 = TEESMC_OS_OPTEE_REVISION_MAJOR;
-	args->a1 = TEESMC_OS_OPTEE_REVISION_MINOR;
+	args->a0 = tee_entry_generic_get_api_call_count() + 3;
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-sunxi/main.c]: "+1" instead of "+3" ?

File: core/arch/arm/plat-stm/conf.mk (hunk 1 above)
Comment (jforissier) [inline core/arch/arm/plat-stm/conf.mk]: Will fix

File: core/arch/arm/plat-sunxi/main.c (hunk 3 above)
Comment (jforissier) [inline core/arch/arm/plat-sunxi/main.c]: There are 3 specific `TEESMC32_*` values handled in https://github.com/jforissier/optee_os/blob/refactoring/core/arch/arm/plat-sunxi/main.c#L136 hence +3. Did I miss something?

Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-sunxi/main.c]: My mistake. Forget about my comment.

**Comment by pascal-brand-st-dev** [comment]
Once the indentation is fixed:

    Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

File: core/include/kernel/tee_common_otp.h [hunk 4]
```diff
@@ -36,10 +36,31 @@ struct tee_hw_unique_key {
...
-/* exposed to let tee_init set the key */
-extern uint8_t hw_key_digest[TEE_SHA256_HASH_SIZE];
+#if defined(CFG_OTP_SUPPORT)
 
 void tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey);
 int tee_otp_get_die_id(uint8_t *buffer, size_t len);
 
+#else
+
+#include <string.h>
```
Comment (jbech-linaro) [inline core/include/kernel/tee_common_otp.h]: Wouldn't this be better to keep where we have the other includes?

**Comment by jbech-linaro** [comment]
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: core/include/kernel/tee_common_otp.h (hunk 4 above)
Comment (jforissier) [inline core/include/kernel/tee_common_otp.h]: I have no strong opinion on this. Assuming we move this up, would you still keep the #ifdef guard, or include string.h unconditionally?
@pascal-brand-st-dev what do you prefer?

Comment (pascal-brand-st-dev) [inline core/include/kernel/tee_common_otp.h]: Unconditionnally makes the code cleaner.

Comment (jforissier) [inline core/include/kernel/tee_common_otp.h]: OK I'll fix that.

**Comment by jforissier** [comment]
@jenswi-linaro are you OK with the current state?

File: core/arch/arm/include/tee/entry.h [hunk 5]
```diff
@@ -25,8 +25,8 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-#ifndef TEE_ENTRY_H
-#define TEE_ENTRY_H
+#ifndef ENTRY_H
```
Comment (jenswi-linaro) [inline core/arch/arm/include/tee/entry.h]: The usual pattern for inclusion guards is to take the name the file is included by (in this case `tee/entry.h`) and create a unique define based on that. That guarantees that the guard is unique.

**Comment by jenswi-linaro** [comment]
@jforissier, one minor remark only.
Have you tested that the pager still works with this?

File: core/arch/arm/include/tee/entry.h (hunk 5 above)
Comment (jforissier) [inline core/arch/arm/include/tee/entry.h]: I thought the convention was to use the file name only. Actually many files in OP-TEE do not follow this recommendation, but I see how it is better. Will fix.

**Comment by jforissier** [comment]
How can I test the pager?
Well commit f7f8b56 breaks the build when `CFG_WITH_PAGER=y`:
```
  LD      out/arm-plat-vexpress/core/tee.elf
arm-linux-gnueabihf-ld: OP-TEE can't fit init part into available physical memory
```
Should I add this (which fixes the link error):
```
diff --git a/core/arch/arm/plat-vexpress/link.mk b/core/arch/arm/plat-vexpress/link.mk
index a798d32..4f008b3 100644
--- a/core/arch/arm/plat-vexpress/link.mk
+++ b/core/arch/arm/plat-vexpress/link.mk
@@ -55,6 +55,7 @@ $(link-out-dir)/rodata_unpaged.ld.S: $(link-out-dir)/unpaged.o
 objs-init-rem += core/arch/arm/tee/arch_svc.o
 objs-init-rem += core/arch/arm/tee/arch_svc_asm.o
 objs-init-rem += core/arch/arm/tee/init.o
+objs-init-rem += core/arch/arm/tee/entry.o
 entries-init += _start
 objs-init := \
        $(filter-out $(addprefix $(out-dir)/, $(objs-init-rem)), $(objs))
```
And then how to test? With `setup_qemu_optee.sh`?

**Comment by jenswi-linaro** [comment]
To test with pager you compile with CFG_WITH_PAGER=y and boot, run a few test or all of them. If there's any problem it usually happens very early.

File: core/arch/arm/include/tee/entry.h (hunk 5 above)
Comment (jenswi-linaro) [inline core/arch/arm/include/tee/entry.h]: Yes, I guess it's partially because we haven't enforced this much, but also because we've renamed many files.

**Comment by jforissier** [comment]
@jenswi-linaro I fixed a link issue and I can confirm it runs xtest successfully with pager enabled.

**Comment by jenswi-linaro** [comment]
Good, then I'm happy with this too. :-)

**Comment by jforissier** [comment]
Commits squashed and review tags applied.

File: core/arch/arm/include/plat_common.h [hunk 6]
```diff
@@ -24,11 +24,11 @@
...
  */
-#ifndef PLAT_TEE_FUNC_H
-#define PLAT_TEE_FUNC_H
+#ifndef PLAT_COMMON_H
+#define PLAT_COMMON_H
 
 #include <kernel/thread.h>
 
-void plat_tee_entry(struct thread_smc_args *args);
+void plat_common_tee_entry(struct thread_smc_args *args);
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/include/plat_common.h]: Sorry not to have seen that before, but this file looks useless.

File: core/arch/arm/plat-stm/main.c [hunk 7]
```diff
@@ -46,8 +46,8 @@
 #include <tee/arch_svc.h>
 #include <console.h>
 #include <asc.h>
-#include <kernel/tee_l2cc_mutex.h>
 #include <assert.h>
+#include <plat_common.h>
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-stm/main.c]: Useless include. Same in other "main.c" files

**Comment by jforissier** [comment]
@pascal-brand-st-dev That's right, plat_common.h is not needed at this point. I have removed it. But it will likely appear again in my next PR ;-)

**Pull request merged**